hail/python/hail/ggplot/geoms.py

```diff
@@ -166,24 +170,43 @@ def _add_legend(self, fig_so_far: go.Figure, aes_name, category, value):
             }
         )
 
-    def apply_to_fig(self, grouped_data, fig_so_far: go.Figure, precomputed, facet_row, facet_col, legend_cache) -> bool:
+    def apply_to_fig(self, grouped_data, fig_so_far: go.Figure, precomputed, facet_row, facet_col, legend_cache, is_faceted) -> bool:
+        traces = []
         legends = {}
         for df in grouped_data:
             values = self._get_aes_values(df)
-            self._add_trace(fig_so_far, df, facet_row, facet_col, values)
+            trace_category = None
             for aes_name in self.aes_legend_groups:
+                category = self._get_aes_value(df, f"{aes_name}_legend")
+                trace_category = category if category is not None else trace_category
                 legends[aes_name] = ({
                     **legends.get(aes_name, {}),
-                    self._get_aes_value(df, f"{aes_name}_legend"): values[aes_name]
+                    category: values[aes_name]
                 })
-
-        number_of_displayed_legends = 0
-        for aes_name, legend_group in legends.items():
-            if len(legend_group) > 1:
-                number_of_displayed_legends += 1
+            traces = [*traces, [fig_so_far, df, facet_row, facet_col, values, trace_category]]
```

Comment:
I'm a bit confused about trace_category. Can you explain to me a bit more about what the goal is here? In particular, I'm generally suspicious of iterating through a group of values and arbitrarily taking the last one that is not None (what if there are multiple not None `category`s?).

Reply:
`trace_category` is what will be used to label this group on the legend iff we are making an interactive plot, so in my example above, if we only had the `color` aesthetic and not the `shape` one, `trace_category` would end up being `"bad"` in one iteration, and `"good"` in the other. Because there's only one `shape` category and it was created by default instead of using a category from a user-provided aesthetic mapping, the value of `category` when we're looking at the `shape` aesthetic in the iteration is `None`. Since `trace_category` will only actually be used if we are making an interactive plot, and since we only support interactivity when there's one or fewer aesthetic mappings that will appear in the legend with multiple values, we can assume that if we have a `category` that is `None`, we have exactly one `category` that is `None`.

Agreed that this is suspicious haha, but hopefully we can ditch plotly and the way we currently do grouped legends soonish 😸

Reply:
OK, so, check that I understand this correctly:

`grouped_data` is `List[DataFrame]` where the DataFrame contains one column for each Hail field needed to render this geometry. I *think* each DataFrame represents a different "grouping" of the data. In particular, I *think* this comes from `Stat.listify`. We should add a type to that method. I think it returns `List[DataFrame]`. I also think there's a bug in StatNone. It should return `[pd.DataFrame({})]`.

IIUC, there's a "grouping" for each "visually manifested" scale. "visually manifested" means it's a discrete scale and not part of the tooltip. I think this should be called `data_by_visible_scale` or something like that.

Alright, so, IIUC, we're trying to collect all the possible values of these visually manifested scales. From that information, we can determine if we need a legend and, furthermore, if we need a static legend.

I think it's also the case that `color_legend` is the description *of the legend itself* for `color`s. Same for shape_legend and shape.

When we're looping through the `data_by_visible_scale`, it's not obvious that we don't get conflicting values for `category` across different data frames. Let's be concrete. In your example above there are four facets ({good, bad} cross {minor, major}) with two scales: color & shape. Color is perfectly correlated with {good, bad}. `data_by_visible_scale` will have a data frame for each combination of color and shape that appears in that particular facet. OK, so after the first iteration of this loop, legends should be this:
```
{
    "shape": {"shape": ["circle"]},
    "color": {"color": ["blue"]}
}
```
Then the next iteration, unless I'm totally misunderstanding here, would blow away these values and replace them with:
```
{
    "shape": {"shape": ["circle", "square"]},
    "color": {"color": ["red"]}
}
```

So clearly I'm misunderstanding something. I think it must be my understanding of `color_legend` / category, what is that?

Reply:
good call, added return type annotations for `listify` and fixed the return value for `StatNone`!

for the example in the PR description, this is what `legends` looks like:
```
### `apply_to_fig` call ###
#### iteration of `grouped_data` loop ####
{'color': {'bad': 'rgb(0, 219, 207)'}, 'shape': {'no': 'square'}}
#### iteration of `grouped_data` loop ####
{'color': {'bad': 'rgb(0, 219, 207)'}, 'shape': {'no': 'square', 'yes': 'circle'}}
### `apply_to_fig` call ###
#### iteration of `grouped_data` loop ####
{'color': {'bad': 'rgb(0, 219, 207)'}, 'shape': {'yes': 'circle'}}
### `apply_to_fig` call ###
#### iteration of `grouped_data` loop ####
{'color': {'good': 'rgb(255, 0, 0)'}, 'shape': {'no': 'square'}}
#### iteration of `grouped_data` loop ####
{'color': {'good': 'rgb(255, 0, 0)'}, 'shape': {'no': 'square', 'yes': 'circle'}}
### `apply_to_fig` call ###
#### iteration of `grouped_data` loop ####
{'color': {'good': 'rgb(255, 0, 0)'}, 'shape': {'yes': 'circle'}}
```

so each aesthetic (shape, color) has a key in legends, which is mapped to a `dict` mapping the possible values for that aesthetic (the "categories" extracted from the corresponding column of the hail `Table` or `Expression`) to their value for that aesthetic.

in the `None` case i mentioned above, we get something like this:
```
{'color': {None: 'black'}, 'shape': {'no': 'circle', 'yes': 'square'}}
```

due to .... some other code somewhere 😭 (maybe somewhere in `ScaleDiscrete`?) ... the value for the category name when we don't actually have more than one category is `None`

does that actually address your question? sorry, i got a bit lost along the way

Reply:
Yeah, I think I kinda see what's happening now. I have a suspicion we could find a slightly simpler, if less efficient, solution by passing over the data twice: once to compute the legend and once to add the trace. No need to act on that now. Let's keep it in mind as we look to Altair though.